GEOS-11038 setting web.xml to servlet version 3

ls91 opened the pull request:
[![GEOS-11038](https://badgen.net/badge/JIRA/GEOS-11038/0052CC)](https://osgeo-org.atlassian.net/browse/GEOS-11038)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

This pull request is to address issue: https://osgeo-org.atlassian.net/browse/GEOS-11038

Setting the version to 3 allows for GeoServer to be deployed under WebLogic 14.1.1

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

mprins reviewed (commented):
web-app 3 is part of Java EE 6, see https://www.oracle.com/webfolder/technetwork/jsc/xml/ns/javaee/index.html#6 so i see no problem there.

Not sure about any possible side effects; there is some code that parses the web.xml, eg. `GeoServerJ2eeRoleService` that might trip up if the web.xml is not schema-valid.

Jetty already uses  3.1
https://github.com/geoserver/geoserver/blob/7636c1dccc9331b8c6e8a7e660602b066b519001/src/release/jetty/etc/webdefault.xml#L3-L7

sikeoka commented:
Since GeoServer is already using 3.1 in some places, why not make it consistent here and use servlet 3.1 instead of 3.0?

aaime commented:
Obvious test failure, the test should be rewritten to perform schema validation instead (and that may end up requiring some changes in the XML contents too):

```
2023-06-21T12:39:05.0120930Z 12:39:05,011 [INFO] Running org.geoserver.web.WebXmlTest
2023-06-21T12:39:05.1688708Z 12:39:05,168 [ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.152 s <<< FAILURE! - in org.geoserver.web.WebXmlTest
2023-06-21T12:39:05.1689378Z 12:39:05,168 [ERROR] testWebXmlDTDCompliance(org.geoserver.web.WebXmlTest)  Time elapsed: 0.149 s  <<< FAILURE!
2023-06-21T12:39:05.1690481Z java.lang.AssertionError
2023-06-21T12:39:05.1691232Z 	at org.geoserver.web.WebXmlTest.testWebXmlDTDCompliance(WebXmlTest.java:27)
```

At a minimum, the new web.xml should be tested live under Jetty 9, Tomcat 9, and as @mprins indicated, the roles extractions from web.xml is still working after the upgrade.

ls91 commented:
I'm currently a bit stumped here, I'm not able to replicate these tests failing locally, I'll keep going though but if anyone has any suggestions I'm all ears. I've switched my JDK to match in case that was a cause but its not replicated it. The only other difference is I'm running Windows.

aaime commented:
If you're referring to the failures in gs-metadata, @NielsCharlier is looking into them, no action is required of it about it.

What is missing here are eventual updates to the GeoServerJ2eeRoleServiceTest and in particular its test files, [web1.xml](https://github.com/geoserver/geoserver/blob/52607dd55d84e037dca577f755c1ae6dbc9cb6ba/src/security/security-tests/src/test/resources/org/geoserver/security/impl/web1.xml) and [web2.xml](https://github.com/geoserver/geoserver/blob/52607dd55d84e037dca577f755c1ae6dbc9cb6ba/src/security/security-tests/src/test/resources/org/geoserver/security/impl/web2.xml), with checks that everything is still working fine after the switch away from the DTD.

ls91 commented:
I've updated those test XMLs and the tests are still passing with the roles being extracted. I also added XSD validation to the test for completeness.

jodygarnett commented on src/web/app/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -1,6 +1,10 @@
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd"
+        metadata-complete="false"
+        version="3.1">
```
This change would make me happy - finally not having the DTD :)

jodygarnett reviewed (approved):
This looks good to me, thank you for the improvement.

jodygarnett commented:
There are problems with gs-metadata so you may need to rebase your PR to get a clean run of tests.

aaime commented:
The changes look good and the build is green.

Does anyone see problems in updating the web.xml file in stable/maintenance as well? (thinking whether to backport or not this PR).

jodygarnett commented:
Please backport

aaime commented:
The backport to `2.22.x` failed:
```
The process '/usr/bin/git' failed with exit code 128
```
<details>
<summary>stderr</summary>

```
error: commit 2a536943f43541e3b17ad227c2e83cecd39c1db2 is a merge but no -m option was given.
fatal: cherry-pick failed

```
</details>

<details>
<summary>stdout</summary>

```
[backport-6964-to-2.22.x 1d33842110] GEOS-11038 setting web.xml to servlet version 3
 Author: Luke South <luke.south@bt.com>
 Date: Wed Jun 21 13:02:38 2023 +0100
 1 file changed, 5 insertions(+), 2 deletions(-)
[backport-6964-to-2.22.x 5a50250352] GEOS-11038 setting web.xml to servlet version 3.1 and fixing unit test
 Author: Luke South <luke.south@bt.com>
 Date: Wed Jun 21 22:50:42 2023 +0100
 8 files changed, 6052 insertions(+), 1041 deletions(-)
 delete mode 100644 src/web/app/src/test/java/org/geoserver/web/web-app_2_3.dtd
 create mode 100644 src/web/app/src/test/resources/javaee_7.xsd
 create mode 100644 src/web/app/src/test/resources/javaee_web_services_client_1_4.xsd
 create mode 100644 src/web/app/src/test/resources/jsp_2_3.xsd
 create mode 100644 src/web/app/src/test/resources/web-app_3_1.xsd
 create mode 100644 src/web/app/src/test/resources/web-common_3_1.xsd
[backport-6964-to-2.22.x 493ea31738] GEOS-11038 updating GeoServerJ2eeRoleServiceTest xmls to servlet version 3.1
 Author: Luke South <luke.south@bt.com>
 Date: Fri Jun 23 21:44:57 2023 +0100
 2 files changed, 12 insertions(+), 4 deletions(-)
[backport-6964-to-2.22.x aa2baebc2e] GEOS-11038 Adding XSD validation to tests
 Author: Luke South <luke.south@bt.com>
 Date: Fri Jun 23 22:46:16 2023 +0100
 6 files changed, 6061 insertions(+)
 create mode 100644 src/security/security-tests/src/test/resources/org/geoserver/security/xsd/javaee_7.xsd
 create mode 100644 src/security/security-tests/src/test/resources/org/geoserver/security/xsd/javaee_web_services_client_1_4.xsd
 create mode 100644 src/security/security-tests/src/test/resources/org/geoserver/security/xsd/jsp_2_3.xsd
 create mode 100644 src/security/security-tests/src/test/resources/org/geoserver/security/xsd/web-app_3_1.xsd
 create mode 100644 src/security/security-tests/src/test/resources/org/geoserver/security/xsd/web-common_3_1.xsd

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-2.22.x 2.22.x
# Navigate to the new working tree
cd .worktrees/backport-2.22.x
# Create a new branch
git switch --create backport-6964-to-2.22.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick 42200e612b37451aeba2dd632c322f1be49be6b5,5f16f12094b614582f27e0bdfd2f0f94421b08bd,1d80d3d444897858894c846f555735f044998cc7,9e2a596e8f8220a9bb0e1fae271d834d910921c6,2a536943f43541e3b17ad227c2e83cecd39c1db2
# Push it to GitHub
git push --set-upstream origin backport-6964-to-2.22.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-2.22.x
```
Then, create a pull request where the `base` branch is `2.22.x` and the `compare`/`head` branch is `backport-6964-to-2.22.x`.

aaime commented:
The backport to `2.23.x` failed:
```
The process '/usr/bin/git' failed with exit code 128
```
<details>
<summary>stderr</summary>

```
error: commit 2a536943f43541e3b17ad227c2e83cecd39c1db2 is a merge but no -m option was given.
fatal: cherry-pick failed

```
</details>

<details>
<summary>stdout</summary>

```
[backport-6964-to-2.23.x 98c9d2e585] GEOS-11038 setting web.xml to servlet version 3
 Author: Luke South <luke.south@bt.com>
 Date: Wed Jun 21 13:02:38 2023 +0100
 1 file changed, 5 insertions(+), 2 deletions(-)
[backport-6964-to-2.23.x 1a76b94329] GEOS-11038 setting web.xml to servlet version 3.1 and fixing unit test
 Author: Luke South <luke.south@bt.com>
 Date: Wed Jun 21 22:50:42 2023 +0100
 8 files changed, 6052 insertions(+), 1041 deletions(-)
 delete mode 100644 src/web/app/src/test/java/org/geoserver/web/web-app_2_3.dtd
 create mode 100644 src/web/app/src/test/resources/javaee_7.xsd
 create mode 100644 src/web/app/src/test/resources/javaee_web_services_client_1_4.xsd
 create mode 100644 src/web/app/src/test/resources/jsp_2_3.xsd
 create mode 100644 src/web/app/src/test/resources/web-app_3_1.xsd
 create mode 100644 src/web/app/src/test/resources/web-common_3_1.xsd
[backport-6964-to-2.23.x f5ee8da651] GEOS-11038 updating GeoServerJ2eeRoleServiceTest xmls to servlet version 3.1
 Author: Luke South <luke.south@bt.com>
 Date: Fri Jun 23 21:44:57 2023 +0100
 2 files changed, 12 insertions(+), 4 deletions(-)
[backport-6964-to-2.23.x 133c1d0001] GEOS-11038 Adding XSD validation to tests
 Author: Luke South <luke.south@bt.com>
 Date: Fri Jun 23 22:46:16 2023 +0100
 6 files changed, 6061 insertions(+)
 create mode 100644 src/security/security-tests/src/test/resources/org/geoserver/security/xsd/javaee_7.xsd
 create mode 100644 src/security/security-tests/src/test/resources/org/geoserver/security/xsd/javaee_web_services_client_1_4.xsd
 create mode 100644 src/security/security-tests/src/test/resources/org/geoserver/security/xsd/jsp_2_3.xsd
 create mode 100644 src/security/security-tests/src/test/resources/org/geoserver/security/xsd/web-app_3_1.xsd
 create mode 100644 src/security/security-tests/src/test/resources/org/geoserver/security/xsd/web-common_3_1.xsd

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-2.23.x 2.23.x
# Navigate to the new working tree
cd .worktrees/backport-2.23.x
# Create a new branch
git switch --create backport-6964-to-2.23.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick 42200e612b37451aeba2dd632c322f1be49be6b5,5f16f12094b614582f27e0bdfd2f0f94421b08bd,1d80d3d444897858894c846f555735f044998cc7,9e2a596e8f8220a9bb0e1fae271d834d910921c6,2a536943f43541e3b17ad227c2e83cecd39c1db2
# Push it to GitHub
git push --set-upstream origin backport-6964-to-2.23.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-2.23.x
```
Then, create a pull request where the `base` branch is `2.23.x` and the `compare`/`head` branch is `backport-6964-to-2.23.x`.
